examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.axes import Axes
+from matplotlib.collections import PolyCollection
+from matplotlib.figure import Figure
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+
+### Parameters for a full simulation loop
+
+# For the full simulation, we need to define some additional parameters. These are the number of experiments to be conducted per run, the batch size, the dimension and the points per dimension.
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 3 if SMOKE_TEST else 7
+BATCH_SIZE = 2
+DIMENSION = 3
+POINTS_PER_DIM = 4
+
+### Setup
+
+# We start by initializing the global `poi_list` where we will store the PI from each recommend iteration:
+
+pi_list: list[torch.Tensor] = []
+
+
+# Then, we define the hook that calculates the PI based on the given recommender, search space and measurements:
+
+
+def check_probability_of_improvement(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+) -> list[torch.Tensor]:
+    """Calculate and store the probability of improvement in poi_list.
+
+    For reasons of numerical stability, the function adds some noise to the calculated
+    probabilities.
+    """
+    if searchspace.type != SearchSpaceType.DISCRETE:
+        raise TypeError(
+            f"{searchspace.type} search spaces are not supported yet. "
+            f"Currently only DISCRETE search spaces are accepted."
+        )
+    train_x = searchspace.transform(measurements)
+    train_y = objective.transform(measurements)
+    acqf = ProbabilityOfImprovement()
+    botorch_acqf = acqf.to_botorch(self.surrogate_model, searchspace, train_x, train_y)
+    comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+    pi = botorch_acqf(comp_rep_tensor)
+    pi = pi + 1e-10 * torch.randn(pi.shape)
+    pi_list.append(pi)
+
+
+# Additionally, we define a function that plots the `poi_list` after all recommend  iterations:
+
+
+def plot_pi(
+    pi_list: list[torch.Tensor],
+    ax: Axes,
+    base_name: str,
+    path: Path = Path("."),
+) -> Figure:
+    """Plot the probability of improvement in 3D."""
+    cmap = plt.get_cmap("viridis")
+
```

Comment:
why this whitespace?

Reply:
Just to seperate different objects/context. But I removed it now.